service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryExtractor.java

```diff
@@ -0,0 +1,228 @@
+            .filter(Optional::isPresent)
+            .map(resource -> (Practitioner) resource.get())
+            .distinct()
+            .collect(Collectors.toList());
+    }
+
+    public static List<Triple<Practitioner, Optional<PractitionerRole>, Optional<Organization>>>
+        extractPractitionerRoleTriples(Bundle bundle, List<Practitioner> practitioners) {
+
+        List<PractitionerRole> practitionerRoles = bundle.getEntry()
+            .stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> resource.getResourceType().equals(ResourceType.PractitionerRole))
+            .map(resource -> (PractitionerRole) resource)
+            .filter(practitionerRole -> referencesPractitioner(practitionerRole, practitioners))
+            .collect(Collectors.toList());
+
+        List<Organization> organizations = practitionerRoles.stream()
+            .map(practitionerRole -> practitionerRole.getOrganization().getReferenceElement())
+            .map(reference -> ResourceExtractor.extractResourceByReference(bundle, reference))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .map(resource -> (Organization) resource)
+            .collect(Collectors.toList());
+
+        return practitioners.stream()
+            .map(practitioner -> combineTriple(practitioner, practitionerRoles, organizations))
+            .collect(Collectors.toList());
+    }
+
+    private static Triple<Practitioner, Optional<PractitionerRole>, Optional<Organization>> combineTriple(Practitioner practitioner,
+        List<PractitionerRole> practitionerRoles, List<Organization> organizations) {
+
+        var practitionerRole = extractRelevantPractitionerRole(practitioner, practitionerRoles);
+        if (practitionerRole.isPresent()) {
+            var organization = extractRelevantOrganization(practitionerRole.get(), organizations);
+            return Triple.of(practitioner, practitionerRole, organization);
+        } else {
+            return Triple.of(practitioner, Optional.empty(), Optional.empty());
+        }
+    }
+
+    public static Optional<Resource> extractObservationResource(Observation observation, Bundle bundle, ResourceType resourceType) {
+        var reference = observation.getPerformer()
+            .stream()
+            .filter(reference1 -> reference1.getReferenceElement().getResourceType().equals(resourceType.name()))
+            .findFirst();
+        return reference.flatMap(value -> ResourceExtractor.extractResourceByReference(bundle, value.getReferenceElement()));
+    }
+
+    private static boolean referencesPractitioner(PractitionerRole practitionerRole, List<Practitioner> practitioners) {
+        var practitionerRolePractitionerId = practitionerRole.getPractitioner().getReferenceElement().getIdPart();
+        return practitioners.stream()
+            .anyMatch(practitioner -> practitioner.getIdElement().getIdPart().equals(practitionerRolePractitionerId));
```

Comment:
For atomisation of operations:
```suggestion
        return practitioners.stream()
            .map(Practitioner::getIdElement)
            .map(IdType::getIdPart)
            .anyMatch(practitionerRolePractitionerId::equals);
```